### PR TITLE
add new optional for bundle validate

### PR DESCRIPTION
**Pull request by camilamacedo86**
**Description of the change:**

- Add a new optional flag to pass optional values to the validators.  For more info see the EP: https://github.com/operator-framework/enhancements/blob/master/enhancements/operator-hub-check.md

Screen Shot 2021-03-24 at 01 43 43
Screen Shot 2021-03-24 at 01 44 06

**Motivation for the change:**

Allow SDK to use the new checks added for OperatorHub.io

**NOTE** This PR is using my fork. We need to get merged first: https://github.com/operator-framework/api/pull/103 and follow-up: https://github.com/operator-framework/api/pull/111

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [x] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [x] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

File: go.mod
Comment (camilamacedo86) [inline go.mod:49]: /hold

It is using my fork. We cannot merge it until we are able to get the changes merged in the api.

**Comment by estroz** [comment]
I'm not sure I see the value in supplying an optional value. Shouldn't the CSV validator easily be able to perform some checks if a CSV value is present, and skip them if not present?

**Comment by estroz** [comment]
For the sake of the argument, lets say this idea is accepted in general.

Lets also say two different validators have the same optional value key, but different semantics and/or value. How does `--optional-values` allow differentiation? The user may have to know about all validators that accept that key.

There are two ways around this:
1. Require optional value keys always have the validator's name as a key prefix, ex. `operatorhub-k8s=1.22`
2. Use some external configuration like [environment variables](https://github.com/operator-framework/api/pull/103#discussion_r606001123) to do (1).

I'd prefer (2) because it is less verbose and does not put the onus of optional value configuration onto operator-sdk, which merely imports and runs validators.

**Comment by camilamacedo86** [comment]
HI @estroz, 

> I'm not sure I see the value in supplying an optional value. Shouldn't the CSV validator easily be able to perform some checks if a CSV value is present, and skip them if not present?

I am not sure if I understand what you mean here. 

The idea of we provides the optional values is to allow us to send optional values to the validators. 
Regarding all needs that we could check so far, we need to pass the k8s/OCP versions to do the checks against the version that we have the intention to publish the project. It is important for the pipeline for example.

**Comment by camilamacedo86** [comment]
Hi @jmrodri, @estroz, 

That would be very helpful if we are able to merge before the release. It will be helpful and collab with our efforts to identify and push the operator authors to use the latest API version. This check is used by the pipeline.

File: testdata/ansible/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml
Comment (camilamacedo86) [inline testdata/ansible/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml:24]: That has no relation with the validator. 
That is just a change in the API.

**Comment by camilamacedo86** [comment]
Hi @estroz, 

Regarding your suggestion in https://github.com/operator-framework/operator-sdk/pull/4693#issuecomment-815252687 that was re-plied https://github.com/operator-framework/operator-sdk/pull/4693#issuecomment-815277896. See that it in fact has no relation with this PR at all. 

This PR, in SDK, only bumps the API lib and add the new flag which allows the users to inform optionally keys/values for the validators. In this way, any way, your suggestion would not fit here and ought to be addressed in the API. 

However, I still not thinking that we should move with that unless we have use case/scenarios that explicitly justify it since it will reduce usability and maintainability. So, in the future, if that be the case we can cange the API with you suggesting and that ought to not affect SDK at all. For now, IMO I think we can address the need as discussed and defined in the EP and implemented so far. 

c/c @estroz @fabianvf

File: testdata/ansible/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml
Comment (estroz) [reply testdata/ansible/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml:24]: It is completely unrelated to the other content in this PR. Break this out into a separate PR.

Comment (camilamacedo86) [reply testdata/ansible/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml:24]: @estroz that is not possible at all. 
These files are updated by the make generate and are a consequence of the bump of the API.

Comment (estroz) [reply testdata/ansible/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml:24]: Ah I see. This is unfortunate since this field does not do anything, nor does the operator-sdk 
"support" this field. I'd like to see it stripped from the resulting manifest bytes before they are written to disk, in a separate PR.

Comment (camilamacedo86) [reply testdata/ansible/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml:24]: Done the bump in another PR: https://github.com/operator-framework/operator-sdk/pull/4750
That is fine. Tks.

Comment (camilamacedo86) [reply testdata/ansible/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml]: Done.

**Comment by camilamacedo86** [comment]
HI @estroz, @jmrodri, @fabianvf  

Now we have here only the validator impl.

**Review by jmrodri — APPROVED** [review]
/lgtm

**Comment by estroz** [comment]
Before this goes in, I would really like to see docs about how this value is used in the api repo, and have them referenced in SDK docs. Otherwise we're just adding some nebulous, obscure feature.

/hold

**Comment by camilamacedo86** [comment]
Hi @marc-obrien, @estroz,   

> Before this goes in, I would really like to see docs about how this value is used in the api repo, and have them referenced in SDK docs. Otherwise, we're just adding some nebulous, obscure feature.

The new option is doc here. See that the CLI commands are doc by default. 
Also, it is documented in the EP: https://github.com/operator-framework/enhancements/blob/master/enhancements/operator-hub-check.md

However, I totally agree that we should have all validators as their options **fully** documented and not only this one. But we have no documentation about that and it never was a blocker for we start to use them in the SDK or their options so far. 

See that doc the validators as their options it is a big task/epic that goes out of scope here and beyond. I raised an issue for we are able to address that. See: https://github.com/operator-framework/api/issues/109

Also, I raised an issue for OLM to add the SDK bundle validation options and doc that properly. We also have no info about that there. See: https://github.com/operator-framework/operator-lifecycle-manager/issues/2097 

And then, it is important to highlight that we cannot add any doc in OLM before this option exists and be released whcih means that if the request here is to do in OLM docs then, that cannot be a blocker either. However, not that I already pushed this info there as well. See https://github.com/operator-framework/olm-docs/pull/144 with the comment that we cannot merge that at all before it is released. https://github.com/operator-framework/olm-docs/pull/144#issuecomment-819015164

**Comment by camilamacedo86** [comment]
Hi @marc-obrien, 

As spoke with @estroz @jmrodri @fabianvf , the new optional-values flag (what is added here is doc via the CLI). Also, how that works is doc in the EP linked in the first comment. 

IHMO:

- we need to document all validators fully as their optional options in the API( https://github.com/operator-framework/api/issues/109)
- we need documentation in SDK that fully describes the operator bundle validation in the section over how to integrate their project with OLM. Currently, in POV that is not clear too. ( more https://github.com/operator-framework/operator-sdk/issues/4528 )
- Then, we can link the API docs to clarify its details and available options according to the tag/version bumped and used in the SDK project.
- Also, we need to add the SDK validation options in the OLM, OCP, partners, internal pipeline, etc docs. we have docs that suggest using the suite test to validate the bundles but that is not in all of them and then, it would be better if we could link when possible the SDK docs in these places instead of duplicating the info. See: https://github.com/operator-framework/operator-lifecycle-manager/issues/2097

In this way, I think that what we could doc so far is done and what eric request here is a big epic/tasks that go beyond this PR scope. See, that we cannot document its options with SDK on the OLM/OCP/partners(certified) docs, etc., before that exists and be released in SDK. 

Also, to clarifies **I am 100% in favor of having all validators as its options fully documented in the API as @estroz requested**, however, it is **not** today and, because of this, I tracked the issue ( https://github.com/operator-framework/api/issues/109 ). I am intended and happy to work on that as a low priority or help who can do it. 

Then, I cannot find any other extra place now to add any further doc about this new small option. However, please let me know where/how that should be added if you have it in mind.  

PS.: In POV, what was done until now shows enough to address the need and add the value that we need ( provide a helper to allow users/pipeline to check if their projects have or not deprecated APIs and will or not work against specific k8s 1.22) as the first step.

**Comment by camilamacedo86** [comment]
Hi @marc-obrien, 

By @estroz 
> Before this goes in, I would really like to see docs about how this value is used in the api repo, and have them referenced in SDK docs. Otherwise we're just adding some nebulous, obscure feature.

By @marc-obrien 
> /hold totally agree with Eric's points

As spoke with @estroz @jmrodri @fabianvf , the new optional-values flag (what is added here is doc via the CLI). Also, how that works is doc in the EP linked in the first comment. 

IHMO:

- we need to document all validators fully as their optional options in the API( https://github.com/operator-framework/api/issues/109)
- we need documentation in SDK that fully describes the operator bundle validation in the section over how to integrate their project with OLM. Currently, in POV that is not clear too. ( more https://github.com/operator-framework/operator-sdk/issues/4528 )
- Then, we can link the API docs to clarify its details and available options according to the tag/version bumped and used in the SDK project.
- Also, we need to add the SDK validation options in the OLM, OCP, partners, internal pipeline, etc docs. we have places that suggested check the bundles via the suite before publishing such as (https://github.com/operator-framework/olm-docs/pull/1440) but that is not in all of them (e.g. https://github.com/operator-framework/operator-lifecycle-manager/issues/2097) and then, it would be better if we could link when possible the SDK docs in these places instead of duplicating the info.

See, that we cannot document its options with SDK on the OLM/OCP/partners(certified) docs, etc., before that exists and be released in SDK. See https://github.com/operator-framework/olm-docs/pull/144#issuecomment-819015164

Also, to clarifies **I am 100% in favor of having all validators as its options fully documented in the API as @estroz requested**, however, it is **not** today and, because of this, I tracked the issue ( https://github.com/operator-framework/api/issues/109 ). I am intended and happy to work on that as a low priority or help who can do it. However, that is a big epic/tasks that in POV goes beyond this PR scope.

Then, I cannot find any other extra place now to add any further doc about this new small option. However, please let me know where/how that should be added if you have it in mind.  

PS.: In POV, what was done until now shows enough to address the need and add the value that we need ( provide a helper to allow users/pipeline to check if their projects have or not deprecated APIs and will or not work against specific k8s 1.22) as the first step. But, we can always work with follow-ups and further improvements when/if required according to use cases and prioritization (agile development).

File: website/content/en/docs/olm-integration/generation.md
Comment (camilamacedo86) [inline website/content/en/docs/olm-integration/generation.md:251]: @estroz @jmrodri @fabianvf @marc-obrien I added a note here as well.

**Comment by camilamacedo86** [comment]
Hi @estroz, @jmrodri, @fabianvf;

- This pr is also doc the new option CLI/bundle validate
- The key k8s was renamed to k8s-version: https://github.com/operator-framework/api/pull/111
- The EP was supplement with the follow up suggestions for optional values ; https://github.com/operator-framework/enhancements/pull/79
- Also we have the PR to doc that in OLM: https://github.com/operator-framework/olm-docs/pull/144/files
- + issues tracked for the further gaps docs found. 

Please, let me know if has anything else that we ought to address.

**Review by estroz — APPROVED** [review]
/lgtm

**Pull request merged**
